pkg/doc/presexch/definition.go

```diff
@@ -0,0 +1,361 @@
+	Purpose    string                   `json:"purpose,omitempty"`
+	Rule       Selection                `json:"rule,omitempty"`
+	Count      *int                     `json:"count,omitempty"`
+	Min        int                      `json:"min,omitempty"`
+	Max        int                      `json:"max,omitempty"`
+	From       string                   `json:"from,omitempty"`
+	FromNested []*SubmissionRequirement `json:"from_nested,omitempty"`
+}
+
+// InputDescriptor input descriptors.
+type InputDescriptor struct {
+	ID          string                 `json:"id,omitempty"`
+	Group       []string               `json:"group,omitempty"`
+	Name        string                 `json:"name,omitempty"`
+	Purpose     string                 `json:"purpose,omitempty"`
+	Metadata    map[string]interface{} `json:"metadata,omitempty"`
+	Schema      []*Schema              `json:"schema,omitempty"`
+	Constraints *Constraints           `json:"constraints,omitempty"`
+}
+
+// Schema input descriptor schema.
+type Schema struct {
+	URI      string `json:"uri,omitempty"`
+	Required bool   `json:"required,omitempty"`
+}
+
+// Holder describes Constraints`s  holder object.
+type Holder struct {
+	FieldID   []string    `json:"field_id,omitempty"`
+	Directive *Preference `json:"directive,omitempty"`
+}
+
+// Constraints describes InputDescriptor`s Constraints field.
+type Constraints struct {
+	LimitDisclosure bool        `json:"limit_disclosure,omitempty"`
+	SubjectIsIssuer *Preference `json:"subject_is_issuer,omitempty"`
+	IsHolder        []*Holder   `json:"is_holder,omitempty"`
+	Fields          []*Field    `json:"fields,omitempty"`
+}
+
+// Field describes Constraints`s Fields field.
+type Field struct {
+	Path      []string    `json:"path,omitempty"`
+	ID        string      `json:"id,omitempty"`
+	Purpose   string      `json:"purpose,omitempty"`
+	Filter    *Filter     `json:"filter,omitempty"`
+	Predicate *Preference `json:"predicate,omitempty"`
+}
+
+// Filter describes filter.
+type Filter struct {
+	Type             *string                `json:"type,omitempty"`
+	Format           string                 `json:"format,omitempty"`
+	Pattern          string                 `json:"pattern,omitempty"`
+	Minimum          StrOrInt               `json:"minimum,omitempty"`
+	Maximum          StrOrInt               `json:"maximum,omitempty"`
+	MinLength        int                    `json:"minLength,omitempty"`
+	MaxLength        int                    `json:"maxLength,omitempty"`
+	ExclusiveMinimum StrOrInt               `json:"exclusiveMinimum,omitempty"`
+	ExclusiveMaximum StrOrInt               `json:"exclusiveMaximum,omitempty"`
+	Const            StrOrInt               `json:"const,omitempty"`
+	Enum             []StrOrInt             `json:"enum,omitempty"`
+	Not              map[string]interface{} `json:"not,omitempty"`
+}
+
+// IsValid validates presentation definition.
+func (pd *PresentationDefinition) IsValid() error {
```

Comment:
Why we are not using JSON schema validator??

Reply:
something like [this](https://github.com/trustbloc/edge-agent/blob/master/cmd/wallet-web/src/pages/chapi/wallet/common/presentationDefSchema.js#L12-L168)

Reply:
I am using schema validator only for tests. It will require marshaling every time you want to validate definitions.
https://github.com/hyperledger/aries-framework-go/pull/2404/files#diff-1a239b0b79a42a8274b0eb84d4be2f453f4b22cf56e0559eb6c65871d8f190feR577

Reply:
In this object, are you getting presentation definition as objects? we follow JSON schema validation pattern in other entities like DIDDoc. It is easier to maintain, faster and caller gets exact validation errors. Are you using the one in RFC for schema validation in tests? They don't cover all validation rules.

Reply:
`Are you using the one in RFC for schema validation in tests?`  - yes
`They don't cover all validation rules.` - true
I can use `schema validator` but as I said it will marshall the structure one more time. @troyronda @llorllale 
See https://github.com/xeipuuv/gojsonschema/blob/master/jsonLoader.go#L304

Reply:
you can reuse the one we wrote [here](https://github.com/hyperledger/aries-framework-go/pull/2404#discussion_r544408482). You can create a follow-up task for that for now.

Reply:
@sudeshrshetty It does not follow the RFC

Reply:
Done. I am using a schema validator.

Reply:
The one I wrote in wallet follows old version of RFC.

Reply:
are all validation rules covered in schema you added?? or there gonna be one more PR for that?

Reply:
I added from RFC if they change schema we need to change it.

Reply:
well, then we gonna miss some validations since RFC schema is not complete. We can use RFC schema for tests like you used earlier and advanced one with validation for realtime validating presentation definition.

Reply:
@soluchok can we rename this method to `ValidateSchema()`? I don't know what `IsValid()` would mean even when it is a valid JSON object for the schema.

Reply:
@sudeshrshetty 

> well, then we gonna miss some validations since RFC schema is not complete. We can use RFC schema for tests like you used earlier and advanced one with validation for realtime validating presentation definition.

Can you push PRs to the P-E repo to fix the schema?

Reply:
@llorllale renamed to `ValidateSchema()`. Opened a PR https://github.com/decentralized-identity/presentation-exchange/pull/164

Reply:
@llorllale sure I can, that sounds like a good idea